### PR TITLE
use createRoot when possible for rendering pin components

### PR DESCRIPTION
**Pull request by oshi97**
This PR uses a dynamic imports with a .catch fallback for when react-dom/client does not exist.

MapboxMap code had to be tweaked due to differences between ReactDOM.render and ReactDOM.createRoot().render(). Namely, createRoot() does not modify the container element unless it already exists on the physical page. This means that we can't call document.creatElement('div'), use createRoot, and THEN attach it to the mapbox map. Instead, we have to attach the div to the map first.

This also seems it make it more difficult to unit test, my guess would be because createRoot does not do anything on the server. For now I use some jest.mocks to ensure the right methods are being called in the right environments.

TEST=manual,auto

storybook can start up (react 17)
test site can display custom pin in both react 17 and 18

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -1,5 +1,6 @@
 {
   "compilerOptions": {
+    "module": "esnext",
```
Comment (oshi97) [inline tsconfig.json:3]: this is needed so vscode doesn't complain about us doing dynamic imports without setting `module` to something like "esnext" or "commonjs", since vscode only knows about this tsconfig and not the cjs and esm ones which already have `module` set

File: src/components/utils/renderToContainer.tsx [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+/**
+ * Renders the given reactElement into the container.
+ * Will use createRoot and root.render over ReactDOM.render when possible.
+ */
+const renderToContainer = async (
+  reactElement: ReactElement,
+  container: HTMLElement
+): Promise<void> => {
+  const render = await rendererPromiseWithFallback;
```
Comment (tatimblin) [inline src/components/utils/renderToContainer.tsx:36]: Is this `await` required because of the dynamic import?

File: src/components/utils/renderToContainer.tsx [hunk 3]
```diff
@@ -0,0 +1,40 @@
+import { ReactElement } from 'react';
+
```
Comment (tatimblin) [inline src/components/utils/renderToContainer.tsx:2]: Can you add a heading comment that this can be removed when we stop supporting react 17

File: tests/components/utils/renderToContainer.test.tsx [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+it('can render a react component into native dom elements', async () => {
+  const createRootRender = jest.fn();
+  const createRootSpy = jest.spyOn(require('react-dom/client'), 'createRoot')
+    .mockImplementation(() => ({ render: createRootRender }));
+  const reactDomRenderSpy = jest.spyOn(require('react-dom'), 'render');
+
+  const container = document.createElement('div');
+
+  await renderToContainer(<span>test me</span>, container);
+  const isReactDom18 = packageJson.devDependencies['react-dom'].match(/18\.\d*\.\d*/);
```
Comment (tatimblin) [inline tests/components/utils/renderToContainer.test.tsx:19]: do you have to change the react version on your machine to test each case?

Comment (oshi97) [reply tests/components/utils/renderToContainer.test.tsx:19]: our ci takes care of that since it runs in all react 16 17 and 18
https://github.com/yext/search-ui-react/blob/main/.github/workflows/run-tests.yml
but if you wanted to test it locally then yeah

File: src/components/utils/renderToContainer.tsx (hunk 3 above)
Comment (oshi97) [reply src/components/utils/renderToContainer.tsx:2]: updated!

File: tests/components/utils/renderToContainer.test.tsx (hunk 4 above)
Comment (tatimblin) [reply tests/components/utils/renderToContainer.test.tsx:19]: ci covering that sounds good to me

**Review by tatimblin — APPROVED** [review]
+1 (percy failed but no snapshot looks broken, maybe rerun?)

File: tests/tsconfig.json [hunk 5]
```diff
@@ -1,5 +1,8 @@
 {
   "extends": "../tsconfig.json",
+  "compilerOptions": {
+    "resolveJsonModule": true
```
Comment (nmanu1) [inline tests/tsconfig.json:4]: why is this needed?

Comment (tatimblin) [reply tests/tsconfig.json:4]: To import the package.json I think

Comment (nmanu1) [reply tests/tsconfig.json:4]: oh right, I see that in the test file now
